Xin add l0

XinWangIntel opened the pull request:
Add L0 draft

XinWangIntel committed:
Add level-zero backend initial, need code format.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Can compile, crash now.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Add eventpool and destructor

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Add new register

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
can run. need fix deallocation

Signed-off-by: xin1.wang <xin1.wang@intel.com>

Commit:
Explicitly define positional order of operands to layer op (#1604)

YangleiZouIntel committed:
Add OpenVINO Gather (#1606)

* Add OpenVINO Gather Op

* Add file

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

YangleiZouIntel committed:
Add buffer transfer optimization for OpenCL (#1590)

* Add createMinimizeBufferTransfersPass
Add createGatherGpuLaunchFuncsPass
Update on gpu_to_comp pass

* Fix mlir tests

* Update Cmake files

* Refactoring

* Add comments

* Refactoring

* Fix

* Make logic clearer

* Fix

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

tzerrell committed:
Cast result branches of select (#1570)

* Cast result branches of select

* Fix merge

* Remove duplicate LogicalAnd

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

YangleiZouIntel committed:
Add buffer transfer optimization for Vulkan (#1605)

* Add createMinimizeBufferTransfersPass
Add createGatherGpuLaunchFuncsPass
Update on gpu_to_comp pass

* Fix mlir tests

* Update Cmake files

* Refactoring

* Add comments

* Refactoring

* Fix

* Make logic clearer

* Vulkan update

* Simplify

* Update goldens due to fix on Vulkan code the more correctly measures time

Co-authored-by: Jeremy Bruestle <jeremy.bruestle@intel.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

XinWangIntel committed:
Can pass single kernel.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

jbruestle committed:
Rewrite hoisting to work on new structure (#1607)

* Rewrite hoisting to work on new structure, enable it, fix bugs in tile-to-pxa and dealloc injection induced by hoisting

* Added cmake stuff (maybe)

* Fix issues with padding not being handled properly on hoisted values

* Clean up debugging messages and fix type

* More attempts at cmake

* Typo in type fix

Commit:
Add support for conditional runtimes (#1608)

* Add support for conditional runtimes

* Add missing defs

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

tzerrell committed:
Add ShapeOf with skipped test (#1571)

* Add ShapeOf with skipped test

* Minor changes to ShapeOf

Co-authored-by: nchoudhu <namrata.choudhury@intel.com>
Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

abialas1 committed:
access qualifier pass (#1611)

haoyouab committed:
add axis parameter and update feature to Scatterop (#1610)

Co-authored-by: mergify[bot] <37929162+mergify[bot]@users.noreply.github.com>

XinWangIntel committed:
Add submit before each wait.
Reset command_list after each submit.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XingHongChenIntel committed:
Add CumSum op (#1609)

* add cum_sum op.

* remove rank ASSERT.

* clean code.

* fix exception message.

* moving exclusive mode to oplib CumSum. and add test.

* modify CumSum op, apply new oplib cumsum.

* update palidml/op/_init_.py.

Co-authored-by: Yanglei Zou <yanglei.zou@intel.com>

Commit:
Fix layer op bugs (#1615)

* Fix layer op bugs

* Turn down logs

Commit:
Print compilation errors to stderr (#1617)

XinWangIntel committed:
Open level-Zero

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Merge branch 'plaidml-v1' into xin-add-L0

XinWangIntel committed:
Add access qualifier pass.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Do zeInit in runtime and reuse queues.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Close skipped test, check group size.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Merge branch 'plaidml-v1' into xin-add-L0

XinWangIntel committed:
Add affinex from 93e59c8b76848b6e9bbf6

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Merge branch 'plaidml-v1' into xin-add-L0

XinWangIntel committed:
Follow intel_gen_ocl_spirv changes.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Code refactor

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Use zeKernelSetGroupSize set local size.
Use zeCommandAppendLaunchKenrel set global size.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Move utils to level_zero lib.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

YangleiZouIntel committed:
Add a missing file
Changed level_zero pass prefix

XinWangIntel committed:
Merge pull request #40 from Flex-plaidml-team/L0

Add a missing file

XinWangIntel commented:
Can pass resnet50 with graphic card(630) and compute-runtime(20.40.18075), but failed with graphic card(650) with compute-runtime( 20.37.0), driver version matters.

XinWangIntel committed:
Add bazel BUILD file.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Open all tests in edsl_test, now cmake version shall test by hand,
bazel version can use skip file to test.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Close Dot in edsl.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Add level_zero BUILD file.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Use lws in setGroupSize, and gws/lws in launchKenrel.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

YangleiZouIntel commented on plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt:
```diff
@@ -0,0 +1,31 @@
+CppEdsl.HigherPrecisionConstants
```
Are all these tests still not pass?

XinWangIntel replied:
Some of them can pass, i have remove them from the file. Others may fail for stdx lowering, opencl spirv namespace(we just use GLSL.450), IR lowering error...

YangleiZouIntel replied:
Thanks, is this level_zero skip list the same as opencl skip list? I assume there is a little difference. Yes most the failed tests are due to spirv ocl ops not added in upstream yet, and the pass loweing to spirv is default to GLSL450 ops.

XinWangIntel replied:
No, they shall be same,  but  some cases in opencl skip list can pass now (such like Max), so i remove them.  Not appropriate to change opencl skip list in this PR.

YangleiZouIntel commented on plaidml/edsl/tests/edsl_test-skip.intel_level_zero.txt:
```diff
@@ -0,0 +1,31 @@
+CppEdsl.InterpolatedGatherCubicDouble
+CppEdsl.InterpolatedGatherCubicHalf
+CppEdsl.InterpolatedGatherMultiDIndices
+CppEdsl.BitNot
```
This one is redundant.

XinWangIntel replied:
Removed

YangleiZouIntel commented on pmlc/BUILD:
```diff
@@ -25,9 +25,11 @@ filegroup(
         "//pmlc/dialect/tile/tests:all",
         "//pmlc/dialect/xsmm/tests:all",
         "//pmlc/rt/opencl/tests:all",
+        "//pmlc/rt/levle_zero/tests:all",
```
Typo: level_zero

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/.gdb_history:
```diff
@@ -0,0 +1,8 @@
+where
```
This file is accidently added.

XinWangIntel replied:
Removed

YangleiZouIntel commented on pmlc/conversion/comp_to_llvm/passes.td:
```diff
@@ -17,4 +17,13 @@ def ConvertCompToLLVM : Pass<"pmlc-convert-comp-to-llvm", "mlir::ModuleOp"> {
   ];
 }
 
+def ConvertCompToLevelZero : Pass<"pmlc-convert-comp-to-level-zero", "mlir::ModuleOp"> {
```
This should be deleted.

XinWangIntel replied:
Removed as we use comp_to_llvm now, but i wonder if we shall keep a separate pass for level zero as we may need some info such as where to add submit and when we can release an event.

YangleiZouIntel replied:
This is a good point. I think we should consider adding a new pass to add submit Ops in the future (and also for vulkan).

YangleiZouIntel replied:
We definitly will have L0 specific passes including the one discussed here to improve performance.

XinWangIntel replied:
If they use comp_to_llvm with prefix for all targets, we may need to add a pass before or after that to add more target specific changes

YangleiZouIntel commented on pmlc/conversion/comp_to_llvm/passes.h:
```diff
@@ -33,6 +33,24 @@ std::unique_ptr<mlir::Pass> createConvertCompToLLVMPass();
 std::unique_ptr<mlir::Pass>
 createConvertCompToLLVMPass(const std::string &prefix);
 
+// ============================================================================
```
This should be deleted.

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_device.cc:
```diff
@@ -0,0 +1,106 @@
+                               const ze_device_handle_t &device,
+                               ze_command_queue_group_properties_t properties)
+    : properties(properties) {
+  queue = lzu::create_command_queue(context, device, 0,
```
You might want to add llvm style comments `/*description*/ 0, ` for these 0

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.cc:
```diff
@@ -0,0 +1,201 @@
+  // Need to find a better way to find right group count.
+  lzu::append_launch_function(list, kernel, &gws, resultE, dependencies.size(),
+                              dependencies.data());
+  // As changes in wrapper, do not need to divide now.
```
You can delete these comments now.

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_device.cc:
```diff
@@ -0,0 +1,106 @@
+}
+
+LevelZeroDevice::~LevelZeroDevice() {
+  // Once multiple devices, shall change
```
This comments needs some rewording and other functions will need changes as well.  
In addition I think we will need one more abstraction level/interface for multi-devices.

XinWangIntel replied:
Removed

YangleiZouIntel replied:
Thanks, do you mean removed the comments here or do you mean something else? please check.

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.cc:
```diff
@@ -0,0 +1,201 @@
+    eventPool.destroy_event(event->getEvent());
+  }
+  // for (size_t i = 0; i < kernels.size(); i++) {
+  //  delete kernels[i];
```
kernels are not used now?

XinWangIntel replied:
comp dialect now create xxx_destroy_kernel to release kernel, so remove these.  But not sure how it will change

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.h:
```diff
@@ -0,0 +1,151 @@
+  lzu::zeEventPool eventPool;
+  // clear list
+  std::vector<std::unique_ptr<LevelZeroEvent>> events;
+  std::vector<LevelZeroKernel *> kernels;
```
`kernels` is not used?

XinWangIntel replied:
Removed

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.h:
```diff
@@ -0,0 +1,151 @@
+  // clear list
+  std::vector<std::unique_ptr<LevelZeroEvent>> events;
+  std::vector<LevelZeroKernel *> kernels;
+  std::vector<LevelZeroMemory *> memories;
```
<del>Maybe it is better to use `std::vector<std::shared_ptr<LevelZeroMemory>>` so we do not need to manually free the memory.

Ok, this vector contains all the memories that explicitly deallocated, I think we can keep it as it is.

XinWangIntel replied:
Want to control the time we release resource, as the lifespan of an object is hard to decided until we have a solid pipeline upon.

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.h:
```diff
@@ -0,0 +1,151 @@
+  std::shared_ptr<LevelZeroDevice> device;
+  LevelZeroQueueUser queueUser;
+  lzu::zeEventPool eventPool;
+  // clear list
```
<del>What does this comment mean?

YangleiZouIntel replied:
Ok, these are resources needs to be released. Maybe reword it to something more clear.

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.cc:
```diff
@@ -0,0 +1,201 @@
+  // Need a way to ensure event is not used and then can release it
+  eventPool.InitEventPool(device->getLevelZeroContext(), 600);
+
+  // May create queue based on properties once multiple device used
```
This comment needs some rewording.

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.cc:
```diff
@@ -0,0 +1,201 @@
+
+  LevelZeroKernel *kernel = new LevelZeroKernel(module, name);
+  // kernels.push_back(kernel);
+  return kernel;
```
`return  new LevelZeroKernel(module, name);`

XinWangIntel replied:
Removed kernels

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_wrappers.cc:
```diff
@@ -0,0 +1,250 @@
+  instance = nullptr;
+}
+/*
+void *levelZeroAlloc(void *invocation, size_t bytes) {
```
Please clean up.

XinWangIntel replied:
Done

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_wrappers.cc:
```diff
@@ -0,0 +1,250 @@
+
+void level_zero_wait(uint32_t count, ...) {
+  // No submit created in IR, so call by hand
+  level_zero_submit(instance);
```
So we are submitting at every comp.Wait Op (is it 1:1 mapping to level_zero_wait?)? Not sure of the overhead, this may be something we should keep in mind when tuning the performance.

XinWangIntel replied:
In L0, each command shall be executed by hand and the command list needs to be reset to be use again,  so do submit before wait to avoid too much submit.  May tune it based on comp dialect.

YangleiZouIntel commented:
Mostly nitpicking. This is great work!

XinWangIntel committed:
Code refactor.
Add pmlc-level-zero-runner.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Merge branch 'plaidml-v1' into xin-add-L0

XinWangIntel committed:
Change code to follow new style.

Make comp_to_llvm.cc use prefix to create symbols.

Handle submit before wait.

Delete kernel when called.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Close lit tests. add new pass to runner, remove soem declarations.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

YangleiZouIntel commented on plaidml/bridge/openvino/tests/functional/shared_tests_instances/single_layer_tests/activation.cpp:
```diff
@@ -47,7 +47,7 @@ const std::map<ActivationTypes, std::vector<std::vector<float>>> activationTypes
     {Mish, {{}}},
     {HSwish, {{}}},
-    {SoftPlus, {{}}}
+    {SoftPlus, {{}}},
```
Are these openvino changes related to L0? I think we should make them a separate PR.

XinWangIntel replied:
This from latest plaidml-v1 branch

YangleiZouIntel replied:
I see.

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_device.cc:
```diff
@@ -0,0 +1,107 @@
+  std::lock_guard<std::mutex> lock(queuesMutex);
+  for (std::unique_ptr<LevelZeroQueueGuard> &guard : queues) {
+    // Use same kind of queue now, open once open new device
+    if (/*!(guard->getLevelZeroProperties() & properties) ||*/ guard
```
Please add a comment or TODO  here, will need to get queue by properties and here we will compare the property struct (including the properties.flag)

XinWangIntel replied:
Add comments, we now use same kind of queue, so skip this check, may open when we find better ways to create a queue

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.cc:
```diff
@@ -0,0 +1,191 @@
+}
+
+LevelZeroInvocation::LevelZeroInvocation(LevelZeroDevice *device)
+    : device{device->shared_from_this()}, queueUser(device->getQueue(p)) {
```
Is the `p` in ` queueUser(device->getQueue(p)` a global vaiable? Can we change it to some default settings as in line 76-78?

YangleiZouIntel replied:
ok, it is a private member now. It is a little wired, but I think it is ok for now and there is enough comments we can improve in the future.

XinWangIntel replied:
Yes, the queue property is not really used now, as we use command queue and command list in L0, we may do complex check when more features are opened.

YangleiZouIntel commented on pmlc/rt/level_zero/level_zero_invocation.cc:
```diff
@@ -0,0 +1,191 @@
+void LevelZeroKernel::enqueue(ze_command_list_handle_t list,
+                              ze_group_count_t gws, ze_group_count_t lws,
+                              ze_event_handle_t &resultE) {
+  // The usage of gws and lws will depend on their meaning.
```
<del>I think this comments is good for us. But for PR to plaidml, we can remove this comment.

YangleiZouIntel replied:
Maybe it is better to replace this comments by the meaning of gws in lzu::append_launch_function

YangleiZouIntel replied:
Please add a new line for unused variable like `(void)lws;` or define a macro like `#define UNUSED(x) [&x]{}()`   then `UNUSED(lws);` to avoid unused warning.

XinWangIntel replied:
Add UNUSED_VARIABLE with (void)(x),  may be save space.     Reword and move comment to enqueueKennel function.

XinWangIntel committed:
Code clean.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Merge branch 'plaidml-v1' into xin-add-L0

XinWangIntel committed:
Code clean.
Add PML_USE_DEVICE_LEVEL_ZERO to inform cmake version to use installed level zero.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Code clean.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Close level_zero_probe_test.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Remove all level zero tests.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
Close LevelZero on CMake build.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

XinWangIntel committed:
CMake will download and build level zero instead of using local version.

Signed-off-by: xin1.wang <xin1.wang@intel.com>

leonling-ll committed:
Add draft L0 profiling

leonling-ll committed:
Detect and handle timestamp counter wrapping

YangleiZouIntel committed:
Fix

YangleiZouIntel committed:
Merge branch 'plaidml-v1' into xin-add-L0

XinWangIntel commented:
Submit to plaidml and will work on that PR.
